coil/src/main/java/dev/chrisbanes/accompanist/coil/Coil.kt

```diff
@@ -0,0 +1,257 @@
+import androidx.ui.graphics.asImageAsset
+import androidx.ui.graphics.painter.ImagePainter
+import androidx.ui.graphics.painter.Painter
+import androidx.ui.unit.IntPx
+import androidx.ui.unit.PxSize
+import coil.Coil
+import coil.request.GetRequest
+import coil.request.GetRequestBuilder
+import coil.request.RequestResult
+import kotlin.math.roundToInt
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+
+private enum class ImageLoadState {
+    Loaded,
+    Empty
+}
+
+private val alpha = FloatPropKey()
+private val brightness = FloatPropKey()
+private val saturation = FloatPropKey()
+
+/**
+ * A composable which loads an image using [Coil] into a [Box], using a crossfade when first
+ * loaded.
+ */
+@Composable
+fun CoilImageWithCrossfade(
+    data: Any,
+    requestBuilder: ((GetRequestBuilder) -> Unit)? = null,
+    alignment: Alignment = Alignment.Center,
+    contentScale: ContentScale = ContentScale.Fit,
+    crossfadeDuration: Int = 1000,
+    modifier: Modifier = Modifier
+) {
+    WithConstraints(modifier) { constraints, _ ->
+        var imgLoadState by stateFor(data) { ImageLoadState.Empty }
+
+        val transitionDef = remember(crossfadeDuration) {
+            transitionDefinition {
+                state(ImageLoadState.Empty) {
+                    this[alpha] = 0f
+                    this[brightness] = 0.8f
+                    this[saturation] = 0f
+                }
+                state(ImageLoadState.Loaded) {
+                    this[alpha] = 1f
+                    this[brightness] = 1f
+                    this[saturation] = 1f
+                }
+
+                transition {
+                    alpha using tween {
+                        duration = crossfadeDuration / 2
+                    }
+                    brightness using tween {
+                        duration = (crossfadeDuration * 0.75f).roundToInt()
+                    }
+                    saturation using tween {
+                        duration = crossfadeDuration
+                    }
+                }
+            }
+        }
+
+        val requestWidth = when {
+            constraints.hasFixedWidth -> constraints.maxWidth
+            constraints.hasBoundedWidth -> constraints.maxWidth
+            else -> constraints.minWidth
+        }
+
+        val requestHeight = when {
+            constraints.hasFixedHeight -> constraints.maxHeight
+            constraints.hasBoundedHeight -> constraints.maxHeight
+            else -> constraints.minHeight
+        }
+
+        val result = GetRequest.Builder(ContextAmbient.current)
+                .data(data)
+                .size(requestWidth.value, requestHeight.value)
+                .apply {
+                    if (requestBuilder != null) {
+                        requestBuilder(this)
+                    }
+                }
+                .build()
+                .executeAsComposable()
```

Comment:
Won't this re-execute on every recompose? Should this be guarded by `onActive(requestWidth, requestHeight)`?

Reply:
Oh wait, `executeAsComposable` guard this. nm

Reply:
Yep, the request will be created every time (there's no way around that while keeping the `requestBuilder` lambda), but the actual execution is guarded. I'll add a comment for this.